### ISSUE TITLE
dead_code lint should not report items within an already reported struct/enum/module

### ISSUE THREAD

**Issue by dhardy** [rust-lang/rust#31188]
For example, a library with this code:

```
mod internal {
    pub struct A;
    impl A {
        pub fn hello() -> &'static str {
            &"hello"
        }
    }
}
```

results in these warnings when compiled:

```
dhardy@localhost:~/small/rust/unused_warnings$ cargo build
Compiling unused_warnings v0.1.0 (file:///home/dhardy/small/rust/unused_warnings)
src/lib.rs:2:5: 2:18 warning: struct is never used: `A`, #[warn(dead_code)] on by default
src/lib.rs:2     pub struct A;
                 ^~~~~~~~~~~~~
src/lib.rs:4:9: 6:10 warning: method is never used: `hello`, #[warn(dead_code)] on by default
src/lib.rs:4         pub fn hello() -> &'static str {
src/lib.rs:5             &"hello"
src/lib.rs:6         }
```

Reporting both struct `A` and method `hello` in this case is noisy and likely misleading (e.g. add the following line to the code):

```
pub use internal::A;
```

In my opinion, items within an already-reported item should not be reported (in this case, only `A` should be reported as unused).

**Comment by djc** [comment]
This seems like a duplicate of #39531.

**Comment by dhardy** [comment]
Yes, seems to be, despite being *older*. Still, the discussion is over there so this one might as well be closed.
